File: packages/java/src/main/java/dev/kreuzberg/ChunkMetadata.java
package dev.kreuzberg;

import com.fasterxml.jackson.annotation.JsonCreator;
import com.fasterxml.jackson.annotation.JsonProperty;
import java.util.Objects;
import java.util.Optional;

/**
 * Metadata describing where a chunk appears within the original document.
 *
 * <p>
 * Includes byte offsets into the content string (UTF-8 boundaries), optional
 * page tracking information, and chunking statistics.
 */
public final class ChunkMetadata {
	private final long byteStart;
	private final long byteEnd;
	private final Long firstPage;
	private final Long lastPage;
	private final Integer tokenCount;
	private final int chunkIndex;
	private final int totalChunks;
	private final HeadingContext headingContext;

	@JsonCreator
	public ChunkMetadata(@JsonProperty("byte_start") long byteStart, @JsonProperty("byte_end") long byteEnd,
			@JsonProperty("first_page") Long firstPage, @JsonProperty("last_page") Long lastPage,
			@JsonProperty("token_count") Integer tokenCount, @JsonProperty("chunk_index") int chunkIndex,
			@JsonProperty("total_chunks") int totalChunks,
			@JsonProperty("heading_context") HeadingContext headingContext) {
		if (byteStart < 0 || byteEnd < byteStart) {
			throw new IllegalArgumentException("Invalid chunk byte range: " + byteStart + "-" + byteEnd);
		}
		if (chunkIndex < 0) {
			throw new IllegalArgumentException("chunkIndex must be non-negative");
		}
		if (totalChunks < 1) {
			throw new IllegalArgumentException("totalChunks must be positive");
		}
		if (firstPage != null && firstPage < 1) {
			throw new IllegalArgumentException("firstPage must be positive");
		}
		if (lastPage != null && lastPage < 1) {
			throw new IllegalArgumentException("lastPage must be positive");
		}
		if (firstPage != null && lastPage != null && lastPage < firstPage) {
			throw new IllegalArgumentException("lastPage must be >= firstPage");
		}
		this.byteStart = byteStart;
		this.byteEnd = byteEnd;
		this.firstPage = firstPage;
		this.lastPage = lastPage;
		this.tokenCount = tokenCount;
		this.chunkIndex = chunkIndex;
		this.totalChunks = totalChunks;
		this.headingContext = headingContext;
	}

	/**
	 * Get the byte offset where this chunk starts in the content string (UTF-8
	 * boundary, inclusive).
	 *
	 * @return start byte offset
	 */
	public long getByteStart() {
		return byteStart;
	}

	/**
	 * Get the byte offset where this chunk ends in the content string (UTF-8
	 * boundary, exclusive).
	 *
	 * @return end byte offset
	 */
	public long getByteEnd() {
		return byteEnd;
	}

	/**
	 * Get the first page number this chunk spans (1-indexed, optional).
	 *
	 * @return first page number, or empty if not tracked
	 */
	public Optional<Long> getFirstPage() {
		return Optional.ofNullable(firstPage);
	}

	/**
	 * Get the last page number this chunk spans (1-indexed, optional).
	 *
	 * <p>
	 * Equal to firstPage for single-page chunks.
	 *
	 * @return last page number, or empty if not tracked
	 */
	public Optional<Long> getLastPage() {
		return Optional.ofNullable(lastPage);
	}

	/**
	 * Get the token count for this chunk (optional).
	 *
	 * @return token count, or empty if not available
	 */
	public Optional<Integer> getTokenCount() {
		return Optional.ofNullable(tokenCount);
	}

	/**
	 * Get the index of this chunk within the total chunks.
	 *
	 * @return zero-based chunk index
	 */
	public int getChunkIndex() {
		return chunkIndex;
	}

	/**
	 * Get the total number of chunks this document was split into.
	 *
	 * @return total chunk count
	 */
	public int getTotalChunks() {
		return totalChunks;
	}

	/**
	 * Get the heading context for this chunk's section (optional).
	 *
	 * @return heading context, or empty if not available
	 */
	public Optional<HeadingContext> getHeadingContext() {
		return Optional.ofNullable(headingContext);
	}

	@Override
	public boolean equals(Object obj) {
		if (this == obj) {
			return true;
		}
		if (!(obj instanceof ChunkMetadata)) {
			return false;
		}
		ChunkMetadata other = (ChunkMetadata) obj;
		return byteStart == other.byteStart && byteEnd == other.byteEnd && Objects.equals(firstPage, other.firstPage)
				&& Objects.equals(lastPage, other.lastPage) && Objects.equals(tokenCount, other.tokenCount)
				&& chunkIndex == other.chunkIndex && totalChunks == other.totalChunks
				&& Objects.equals(headingContext, other.headingContext);
	}

	@Override
	public int hashCode() {
		return Objects.hash(byteStart, byteEnd, firstPage, lastPage, tokenCount, chunkIndex, totalChunks,
				headingContext);
	}

	@Override
	public String toString() {
		return "ChunkMetadata{" + "byteStart=" + byteStart + ", byteEnd=" + byteEnd + ", firstPage=" + firstPage
				+ ", lastPage=" + lastPage + ", tokenCount=" + tokenCount + ", chunkIndex=" + chunkIndex
				+ ", totalChunks=" + totalChunks + ", headingContext=" + headingContext + '}';
	}
}

File: packages/php/src/Types/ChunkMetadata.php
<?php

declare(strict_types=1);

namespace Kreuzberg\Types;

/**
 * Chunk metadata describing offsets within the original document.
 *
 * @property-read int $byteStart Starting byte offset
 * @property-read int $byteEnd Ending byte offset
 * @property-read int|null $tokenCount Number of tokens in chunk
 * @property-read int $chunkIndex Chunk index (0-based)
 * @property-read int $totalChunks Total number of chunks
 * @property-read int|null $firstPage First page number in chunk
 * @property-read int|null $lastPage Last page number in chunk
 * @property-read HeadingContext|null $headingContext Heading hierarchy for this chunk's section
 */
readonly class ChunkMetadata
{
    public function __construct(
        public int $byteStart,
        public int $byteEnd,
        public ?int $tokenCount,
        public int $chunkIndex,
        public int $totalChunks,
        public ?int $firstPage = null,
        public ?int $lastPage = null,
        public ?HeadingContext $headingContext = null,
    ) {
    }

    /**
     * @param array<string, mixed> $data
     */
    public static function fromArray(array $data): self
    {
        /** @var int $byteStart */
        $byteStart = $data['byte_start'] ?? 0;

        /** @var int $byteEnd */
        $byteEnd = $data['byte_end'] ?? 0;

        /** @var int|null $tokenCount */
        $tokenCount = $data['token_count'] ?? null;

        /** @var int $chunkIndex */
        $chunkIndex = $data['chunk_index'] ?? 0;

        /** @var int $totalChunks */
        $totalChunks = $data['total_chunks'] ?? 0;

        /** @var int|null $firstPage */
        $firstPage = $data['first_page'] ?? null;

        /** @var int|null $lastPage */
        $lastPage = $data['last_page'] ?? null;

        /** @var array<string, mixed>|null $rawHeadingContext */
        $rawHeadingContext = is_array($data['heading_context'] ?? null) ? $data['heading_context'] : null;
        $headingContext = $rawHeadingContext !== null
            ? HeadingContext::fromArray($rawHeadingContext)
            : null;

        return new self(
            byteStart: $byteStart,
            byteEnd: $byteEnd,
            tokenCount: $tokenCount,
            chunkIndex: $chunkIndex,
            totalChunks: $totalChunks,
            firstPage: $firstPage,
            lastPage: $lastPage,
            headingContext: $headingContext,
        );
    }
}
